README.md
README
================
Helmut Schütz

  - [PowerTOST](#powertost)
      - [Supported Designs](#supported-designs)
      - [Purpose](#purpose)
      - [Supported](#supported)
          - [Power and Sample Size](#power-and-sample-size)
          - [Methods](#methods)
          - [Helpers](#helpers)
      - [Defaults](#defaults)
          - [Average Bioequivalence](#average-bioequivalence)
          - [Reference-Scaled Average
            Bioequivalence](#reference-scaled-average-bioequivalence)
          - [Dose-Proportionality](#dose-proportionality)
          - [Power Analysis](#power-analysis)
      - [Examples](#examples)
          - [Parallel Design](#parallel-design)
          - [Crossover Design](#crossover-design)
          - [Replicate Designs](#replicate-designs)
          - [Dose-Proportionality](#dose-proportionality-1)
          - [Power Analysis](#power-analysis-1)
          - [Speed Comparisons](#speed-comparisons)
      - [Installation](#installation)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# PowerTOST

The package contains functions to calculate power and estimate sample
size for various study designs used in (not only bio-) equivalence
studies.  
Built 2019-08-16 with R 3.6.1.

## Supported Designs

    #>    design                        name    df
    #>  parallel           2 parallel groups   n-2
    #>       2x2               2x2 crossover   n-2
    #>     2x2x2             2x2x2 crossover   n-2
    #>       3x3               3x3 crossover 2*n-4
    #>     3x6x3             3x6x3 crossover 2*n-4
    #>       4x4               4x4 crossover 3*n-6
    #>     2x2x3   2x2x3 replicate crossover 2*n-3
    #>     2x2x4   2x2x4 replicate crossover 3*n-4
    #>     2x4x4   2x4x4 replicate crossover 3*n-4
    #>     2x3x3   partial replicate (2x3x3) 2*n-3
    #>     2x4x2            Balaam's (2x4x2)   n-2
    #>    2x2x2r Liu's 2x2x2 repeated x-over 3*n-2
    #>    paired                paired means   n-1

Although some replicate designs are more ‘popular’ than others, sample
size estimations are valid for *all* of the following designs:

| design  |  type   | sequences       |
| :-----: | :-----: | --------------- |
| `2x2x4` |  full   | TRTR / RTRT     |
| `2x2x4` |  full   | TRRT / RTTR     |
| `2x2x4` |  full   | TTRR / RRTT     |
| `2x2x3` |  full   | TRT / RTR       |
| `2x2x3` |  full   | TRT / RTR       |
| `2x3x3` | partial | TRR / RTR / RRT |

## Purpose

For various methods power can be *calculated* based on

  - nominal *α*, coefficient of variation (*CV*), deviation of test from
    reference (*θ*<sub>0</sub>), acceptance limits {*θ*<sub>1</sub>,
    *θ*<sub>2</sub>}, sample size (*n*), and design.

For all methods the sample size can be *estimated* based on

  - nominal *α*, coefficient of variation (*CV*), deviation of test from
    reference (*θ*<sub>0</sub>), acceptance limits {*θ*<sub>1</sub>,
    *θ*<sub>2</sub>}, target power, and design.

## Supported

### Power and Sample Size

Power covers balanced as well as unbalanced sequences in crossover or
replicate designs and equal/unequal group sizes in two-group parallel
designs. Sample sizes are always rounded up to achieve balanced
sequences or equal group sizes.

  - Average Bioequivalence (with arbitrary *fixed* limits).
  - Two simultaneous TOST procedures.
  - Non-inferiority *t*-test.
  - Ratio of two means with normally distributed data on the original
    scale based on Fieller’s (‘fiducial’) confidence interval.
  - ‘Expected’ power in case of uncertain (estimated) variability and/or
    uncertain *θ*<sub>0</sub>.
  - Reference-scaled bioequivalence based on simulations.
      - EMA: Average Bioequivalence with Expanding Limits (ABEL).  
      - FDA: Reference-scaled Average Bioequivalence (RSABE) for Highly
        Variable Drugs / Drug Products and Narrow Therapeutic Index
        Drugs (NTIDs).  
  - Iteratively adjust *α* to control the type I error in ABEL and
    RSABE.
  - Dose-Proportionality using the power model.

### Methods

  - Exact
      - Owen’s Q.
      - Direct integration of the bivariate non-central
        *t*-distribution.
  - Approximations
      - Non-central *t*-distribution.
      - ‘Shifted’ central *t*-distribution.

### Helpers

  - Calculate *CV* from *MSE* or *SE* (and vice versa).
  - Calculate *CV* from given confidence interval.
  - Calculate *CV<sub>wR</sub>* from the upper expanded limit of an ABEL
    study.
  - Confidence interval of *CV*.
  - Pool *CV* from several studies.
  - Confidence interval for given *α*, *CV*, point estimate, sample
    size, and design.
  - Calculate *CV<sub>wT</sub>* and *CV<sub>wR</sub>* from a (pooled)
    *CV<sub>w</sub>* assuming a ratio of intra-subject variances.
  - *p*-values of the TOST procedure.
  - Analysis tool for exploration/visualization of the impact of
    expected values (*CV*, *θ*<sub>0</sub>, reduced sample size due to
    dropouts) on power of BE decision.
  - Construct design matrices of incomplete block designs.

## Defaults

  - *α* 0.05, {*θ*<sub>1</sub>, *θ*<sub>2</sub>} (0.80, 1.25). Details
    of the sample size search (and the regulatory settings in
    reference-scaled average bioequivalence) are printed.
  - Note: In all functions values have to be given as ratios, not in
    percent.

### Average Bioequivalence

*θ*<sub>0</sub> 0.95, target power 0.80, design "2x2" (TR|RT), exact
method (Owen’s Q).

### Reference-Scaled Average Bioequivalence

*α* 0.05, point estimate constraint (0.80, 1.25), homoscedasticity
(*CV<sub>wT</sup>* = *CV<sub>wR</sup>*), scaling is based on
*CV<sub>wR</sub>*, target power 0.80, design "2x3x3" (TRR|RTR|RRT),
approximation by the non-central *t*-distribution, 100,000 simulations.

  - EMA, WHO, Health Canada, and many others: Average bioequivalence
    with expanding limits (ABEL).
  - FDA: RSABE.

#### Highly Variable Drugs / Drug Products

*θ*<sub>0</sub> 0.90.

###### EMA

Regulatory constant `0.76`, upper cap of scaling at *CV<sub>wR</sup>*
50%, evaluation by ANOVA.

###### Health Canada

Regulatory constant `0.76`, upper cap of scaling at *CV<sub>wR</sup>*
\~57.4%, evaluation by intra-subject contrasts.

###### FDA

Regulatory constant `log(1.25)/0.25`, linearized scaled ABE (Howe’s
approximation).

#### Narrow Therapeutic Index Drugs (FDA)

*θ*<sub>0</sub> 0.975, regulatory constant `log(1.11111)/0.1`, upper cap
of scaling at *CV<sub>wR</sup>* \~21.4%, design "2x2x4" (TRTR|RTRT),
linearized scaled ABE (Howe’s approximation), upper limit of the
confidence interval of *s<sub>wT</sup>*/*s<sub>wR</sup>* ≤2.5.

### Dose-Proportionality

*β*<sub>0</sub> (slope) `1+log(0.95)/log(rd)` where `rd` is the ratio of
the highest and lowest dose, target power 0.80, crossover design,
details of the sample size search suppressed.

### Power Analysis

Minimum acceptable power 0.70. *θ*<sub>0</sub>, design, conditions, and
sample size method depend on defaults of the respective approaches (ABE,
ABEL, RSABE, NTID).

## Examples

Before running the examples attach the library.

``` r
library(PowerTOST)
```

If not noted otherwise, defaults are employed.

### Parallel Design

Power for total *CV* 0.35, *θ*<sub>0</sub> 0.95, group sizes 52 and 49,
design "parallel".

``` r
power.TOST(CV = 0.35, theta0 = 0.95, n = c(52, 49), design = "parallel")
#> [1] 0.8011186
```

### Crossover Design

Sample size for assumed intra-subject *CV* 0.20.

``` r
sampleN.TOST(CV = 0.20)
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2 crossover 
#> log-transformed data (multiplicative model)
#> 
#> alpha = 0.05, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.95,  CV = 0.2
#> 
#> Sample size (total)
#>  n     power
#> 20   0.834680
```

Sample size for equivalence of the ratio of two means with normality on
original scale based on Fieller’s (‘fiducial’) confidence interval.
*CV<sub>w</sub>* 0.20, *CV<sub>b</sub>* 0.40.  
Note the default *α* 0.025 (95% CI) of this function because it is
intended for studies with clinical endpoints.

``` r
sampleN.RatioF(CV = 0.20, CVb = 0.40)
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>     based on Fieller's confidence interval
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2 crossover
#> Ratio of means with normality on original scale
#> alpha = 0.025, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.95,  CVw = 0.2,  CVb = 0.4
#> 
#> Sample size
#>  n     power
#> 28   0.807774
```

### Replicate Designs

#### ABE

Sample size for assumed intra-subject *CV* 0.45, *θ*<sub>0</sub> 0.90,
3-period full replicate design "2x2x3" (TRT|RTR).

``` r
sampleN.TOST(CV = 0.45, theta0 = 0.90, design = "2x2x3")
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2x3 replicate crossover 
#> log-transformed data (multiplicative model)
#> 
#> alpha = 0.05, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.9,  CV = 0.45
#> 
#> Sample size (total)
#>  n     power
#> 124   0.800125
```

Note that the conventional model assumes homoscedasticity. For
heteroscedasticity we can ‘switch off’ all conditions of one of the
methods for reference-scaled ABE. We assume a σ<sup>2</sup> ratio of ⅔
(*i.e.*, T has a lower variability than R). Only relevant columns of the
data.frame shown.

``` r
reg <- reg_const("USER", r_const = NA, CVswitch = Inf,
                 CVcap = Inf, pe_constr = FALSE)
CV  <- CVp2CV(CV = 0.45, ratio = 2/3)
res <- sampleN.scABEL(CV=CV, design = "2x2x3", regulator = reg,
                      details = FALSE, print = FALSE)
print(res[c(3:4, 8:9)], digits = 5, row.names = FALSE)
#>    CVwT    CVwR Sample size Achieved power
#>  0.3987 0.49767         126         0.8052
```

Similar sample size because the pooled *CV* is still 0.45.

#### ABEL

Sample size assuming homoscedasticity (*CV<sub>w</sub>* = 0.45).

``` r
sampleN.scABEL(CV = 0.45, details = TRUE)
#> 
#> +++++++++++ scaled (widened) ABEL +++++++++++
#>             Sample size estimation
#>    (simulation based on ANOVA evaluation)
#> ---------------------------------------------
#> Study design: 2x3x3 (TRT|RTR) 
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.45; CVw(R) = 0.45
#> True ratio = 0.9
#> ABE limits / PE constraint = 0.8 ... 1.25 
#> EMA regulatory settings
#> - CVswitch            = 0.3 
#> - cap on scABEL if CVw(R) > 0.5
#> - regulatory constant = 0.76 
#> - pe constraint applied
#> 
#> 
#> Sample size search
#>  n     power
#> 36   0.7755 
#> 39   0.8059
```

#### RSABE

#### HVD(P)s

Sample size for a four-period full replicate study (TRTR|RTRT) assuming
heteroscedasticity (*CV<sub>wT</sub>* 0.40, *CV<sub>wR</sub>* 0.50).
Details of the sample size search suppressed.

``` r
sampleN.RSABE(CV = c(0.40, 0.50), design = "2x2x4", details = FALSE)
#> 
#> ++++++++ Reference scaled ABE crit. +++++++++
#>            Sample size estimation
#> ---------------------------------------------
#> Study design: 2x2x4 (TRTR|RTRT) 
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.4; CVw(R) = 0.5
#> True ratio = 0.9
#> ABE limits / PE constraints = 0.8 ... 1.25 
#> Regulatory settings: FDA 
#> 
#> Sample size
#>  n    power
#> 20   0.81509
```

#### NTIDs

Sample size assuming heteroscedasticity (*CV<sub>w</sub>* 0.125,
σ<sup>2</sup> ratio 2.5, *i.e.*, T has a substantially higher
variability than R). Assess additionally which one of the three
components (scaled, ABE, *s<sub>wT</sub>*/*s<sub>wR</sub>* ratio) drives
the sample size.

``` r
CV <- signif(CVp2CV(CV = 0.125, ratio = 2.5), 4)

NEWS
Version 1.4.7.9000
    - Smaller font in 4th screen of pa.X() for NTIDFDA (more lines required).
    - New: README.md (knitted from README.Rmd).
    - Broken FDA URL in many man pages corrected.
    - Bug fix in expsampleN.noninf() wrt operator && with vector arguments 
      (new check in R 3.6.0).
    - Bug in CI.RatioF() fixed. Thanks to Michael (mittyri).
    - Function to simulate subject data & evaluate via models with group effect
      implemented. not yet public.
    - Unified Code base for the scaled ABE functions.
      
Version 1.4-7
2018-04-12 (dedicated to the birthday of Alfie Schuetz :-))
    - Bug corrected in functions using expected power
    - Function type1error.2TOST() no longer available since it suffers from
      insufficient precision to obtain the type 1 error (TIE) via simulations.
      Due to the intersection-union principle the TIE is always upper bounded
      to alpha by theory.
    - Function power.2TOST() now based on simulations to obtain the power 
      of 2 TOSTs (statistical flaw in 4-dimensional t-distribution approach).
    - sampleN.2TOST and man page adapted according to this change.
    - regulator="FDA" implemented in scABEL.ad().
    - New function power.RSABE2L.sds() which implements the 'exact' based
      method for RSABE (ncTOST) of the two Laszlos.
      Documentation included.

Version 1.4-6
Released 2017-08-17
    - Bug fixed in functions relying on simulations if nsims>1e7.
    - Functions CVwRfromU() / U2CVwR() to calculate CVwR from the upper expanded 
      limit of an ABEL study according to EMA's or Health Canada's rules.
    - Power and sample size for TOST: Argument alpha restricted to scalar.
      Internal functions now allow length = 2 (different alpha's
      for the two null hypotheses).
    - Man pages for non-inferiority again updated.
    - Deprecated argument dfCV in expected power functions now completely removed.
    - Data lazy loading (allows access by the name)

Version 1.4-5
Released 2017-05-19
    - Man pages for non-inferiority functions updated to include a
      description of the underlying hypotheses.
    - Remaining man pages updated where term 'Null (true) ratio' was
      mentioned instead of 'True ratio'.
    - Two new test scrips in /test subdirectory added: test_ABEL.R and
      test_RSABE which recalculate the sample size tables given in 
      Tothfalusi & Endrenyi "Sample Sizes for Designing Bioequivalence Studies 
      for Highly Variable Drugs". Contributed by Helmut.
    - power.scABEL.sdsims() now has an argument design_dta to specify the
      design via a data.frame. May be useful for considering missing data.
      Attention! This feature is experimental because the data.frame is not 
      checked at moment.
    - Bug in power.scABEL.sdsims() fixed which gave power=NA if alpha=0.
    - Internal change of coding in design helpers (repeated creation of 
      data.frame with design characteristics avoided).
    - Typo corrected in Expected_Power_for_TOST.pdf.
    - Updated functions scABEL.ad() and sampleN.scABEL.ad() to allow
      subject data simulations via power.scABEL.sdsims() if regulator = "EMA".
      Removed regulator = "ANVISA".
      Changed the order of sequences to be consistent with the other
      functions of PowerTOST.
    - Renamed the variable 'adj. alpha' in sampleN.scABEL.ad() to 'alpha.adj'
      for consistency with scABEL.ad(). Bug corrected if pre-specified
      alpha gives TIE <0.05.
    - Removed regulator = "ANVISA" from pa.scABE().
    - Small bug in the S3 method plot for class 'pwrA' corrected (text
      in the 4th box was truncated on top if saved as a PDF).
      Replaced text() by TeachingDemos::shadowtext() to enhance legibility
      (interfered with underlying grid).

Version 1.4-4
Released 2017-03-15
    - Administrative change for expected power: cubature::adaptIntegrate 
      replaced by cubature::hcubature to reflect change of function name within
      package cubature.
    - Added function power.scABEL.sdsims() to calculate power for the BE decision 
      via scaled (widened) BE acceptance limits (EMA recommended) based on 
      subject data simulations.
    - Low CV (cosmetic) correction in power.scABEL() introduced.
    - Removed default CV = 0.3 from scABEL.ad() and sampleN.scABEL.ad().
      Stops execution if CV not specified.
    - DOI for references added in many man pages. THX to Helmut.
    - Bens description of expected power added in /doc subdirectory
    - BE_power_sample_size_excerpt.pdf updated to reflect the changes in
      computation of OwensQ.
    - Use of removed functions power.scABEL2() and sampleN.scABEL2() in
      pa.scABE() corrected.

Version 1.4-3
Released 2016-11-01
    - Little bug in power.TOST() removed which caused power <0 for large degrees
      of freedom.
    - Functions for expected power of the TOST procedure and sample size based
      on expected power reworked to deal also with uncertainty of theta0 or
      dealing with both uncertainties of CV and theta0.
      Contributed mainly by Benjamin Lang.
    - Functions for expected power of non-inferiority testing and sample size
      based on that expected power reworked to deal also with uncertainty of
      theta0 or dealing with both uncertainties of CV and theta0.
      Contributed mainly by Benjamin Lang.
    - Deprecated functions power.scABEL2() and sampleN.scABEL2() now removed.
      regulator = "ANVISA" no longer allowed in the scaled ABEL functions.
    - OwensQ() simpler/faster implemented. Now based soleyly on numerical
      integrataion in combination with non-central t.
    - Owen's T-function, used in OwensQOwen() now based on algorithm AS76 and
      remarks to that algorithm to avoid numerical errors of the implementation
      via integrate(). 
    - Data.frame ct9.6.6 in data("data2x2x3") added which was missing a long
      time ago.

scABEL.ad.R
#-----------------------------------------------------------------------
# Iteratively adjust alpha to maintain the TIE <= nominal alpha for
# partial and full replicate design and scaled ABE via simulated
# (empirical) power
#
# Author: Helmut Schuetz
#-----------------------------------------------------------------------
scABEL.ad <-function(alpha = 0.05, theta0, theta1, theta2, CV = 0.3,
                     design = c("2x3x3", "2x2x4", "2x2x3"), regulator, n, 
                     alpha.pre = 0.05, imax = 100, tol, print = TRUE, 
                     details = FALSE, setseed = TRUE, nsims = 1e6)
{
  ## Arguments:
  ##   alpha      Nominal alpha (in BE generally fixed to 0.05).
  ##              Lower value only if needed (e.g. to correct for
  ##              multiplicity).
  ##   theta0     If given, power is estimated for the expected GMR.
  ##              Note that the default is 0.90 (different from
  ##              sampleN.scABEL(), where the default is 0.95!
  ##   theta1     Lower margin. Defaults to 0.8.
  ##   theta2     Upper margin. Defaults to 1/theta1.
  ##   CV         Intra-subject CV(s) obtained in a replicate design.
  ##              (ratio, /not/ percent).
  ##              If given as a scalar, the CV of R.
  ##              If given as a vector, CV[1] /must/ be the CV of T and
  ##              CV[2] the CV of R. Important!
  ##              Defaults to 0.3 (maximum TIE for EMA).
  ##   design     "2x2x4", "2x2x3", "2x3x3". Defaults to "2x3x3".
  ##   regulator  "EMA" or "ANVISA". ANVISA recently adopted EMA's rules.
  ##   n          Total sample size or a vector of subjects/sequences.
  ##   nsims      Simulations for the TIE. Should not be <1e6.
  ##   imax       Max. steps in sample size search.
  ##   tol        Desired accuracy (convergence tolerance of uniroot);
  ##              defaults to 1e-6.
  ##   print      Boolean (FALSE returns a list of results).
  ##   details    Boolean (runtime, number of simulations).
  ##   alpha.pre  Pre-specified level.
  ##   setseed    Boolean (default TRUE uses set.seed(123456)).
  ## Returns:
  ##   alpha.adj  Iteratively adjusted alpha which does not inflate the
  ##              TIE (for given CVwR and n).
  ##   CI.adj     The adjusted confidence interval in percent, where
  ##              CI.adj = 100(1-2*alpha.adj).
  ##   TIE.unadj  The empiric Type I Error based on nominal alpha.
  ##   TIE.adj    TIE based on adjusted alpha.
  ##   rel.change Relative change in risk (%) compared to nominal alpha.
  ##   If theta0 is given:
  ##   pwr.unadj  Power for alpha (or, if given, alpha.pre).
  ##   pwr.adj    Power for adjusted alpha.
  ##   rel.loss   Relative loss in power if the sample size was planned
  ##              for alpha and will be evaluated with alpha.adj,
  ##              where rel.loss = 100(pwr.adj - pwr.unadj)/pwr.unadj
  ##   If alpha.pre is given:
  ##   Assessment of TIE; alpha.pre is justified if not > alpha.
  ################################################################
  ## Tested on Win 7 Pro SP1 64bit                              ##
  ##   R 3.3.1 64bit (2016-06-21), PowerTOST 1.4-1 (2016-06-14) ##
  ################################################################
  env <- as.character(Sys.info()[1]) # get info about the OS
  if ((env == "Windows") || (env == "Darwin")) flushable <- TRUE
    else flushable <- FALSE # supress flushing on other OS's
  # acceptance range defaults
  if (missing(theta1) && missing(theta2)) theta1 <- 0.8
  if (missing(theta2)) theta2 = 1/theta1
  # check theta0
  if (missing(theta0)) theta0 <- 0.9
  if (theta0 < theta1 || theta0 > theta2)
    stop("theta0 must be within [theta1, theta2]")
  # check regulator arg
  if(missing(regulator)) regulator <- "EMA"
  # should we also allow "FDA"?
  reg <- reg_check(regulator, choices=c("EMA", "HC", "ANVISA"))
  # set iteration tolerance for uniroot().
  if (missing(tol)) tol <- 1e-6
  design <- match.arg(design)
  CVwT <- CV[1]
  if (length(CV) == 2) CVwR <- CV[2] else CVwR <- CVwT
  no <- 0 # simulation counter
  if (details) ptm <- proc.time()
  if (missing(n) || is.na(n)) {
    if (is.na(alpha.pre) || (alpha.pre != alpha)) {
      al <- alpha.pre    # If pre-specified, use alpha.pre
    } else {
      al <- alpha        # If not, use alpha (commonly 0.05)
    }
    # sample size for targetpower 0.8
    # what about theta1, theta2?
    n <- sampleN.scABEL(alpha = al, CV = CV, theta0 = theta0,
                        design = design, regulator = reg,
                        imax=imax, print = FALSE, details = FALSE,
                        setseed = setseed)[["Sample size"]]
    if (is.na(n))
      stop(paste0("Sample size search in sampleN.scABEL() failed.\n",
                  "Restart with an explicit high n (>1000)."))
    # DL: next should be outside I think. Or?
    if (sum(n) < 6) stop("Sample size too low.")
    no <- 1e5
  }
  if (alpha.pre > alpha) {
    warning(paste0("alpha.pre > alpha doesn't make sense.\n",
                   "alpha.pre was set to alpha."))
    alpha.pre <- alpha
  }
  seqs <- as.numeric(substr(design, 3, 3)) # subjects / sequence
  if (length(n) == 1) n <- nvec(n, seqs)   # vectorize n

  # here we go!
  TIE <- pwr <- rep(NA, 2) # initialize vectors: TIE and pwr
  alpha.adj <- NA          # adjusted alpha
  opt <- function(x) power.scABEL(alpha = x, CV = CV, theta0 = U, n = n,
                                  regulator = reg, design = design,
                                  nsims = nsims, setseed = setseed) - alpha
  # Finds adjusted alpha which gives TIE as close as possible to alpha.
  sig  <- binom.test(x = round(alpha*nsims, 0), n = nsims,
                     alternative = "less",
                     conf.level = 1 - alpha)$conf.int[2]
  method <- "ABE"
  if (CVwR > reg$CVswitch) method <- "ABEL"
  U <- scABEL(CV = CVwR, regulator = reg)[["upper"]]
  # Simulate at the upper (expanded) limit. For CVwR 30% that is
  # 1.25. Due to the symmetry simulations at the lower limit (0.8)
  # would work as well.
  if (alpha.pre != alpha) {
    al <- alpha.pre # If pre-specified, use alpha.pre.
  } else {
    al <- alpha     # If not, use alpha (commonly 0.05).
  }
  designs <- c("2x2x4", "2x2x3", "2x3x3")
  type    <- c("RTRT|TRTR", "RTR|TRT", "RRT|RTR|TRR") # clear words
  if (print) { # Show input to keep the spirits of the user high.
    cat("\n+++++++++++ scaled (widened) ABEL +++++++++++\n")
    cat("         iteratively adjusted alpha\n")
    cat("---------------------------------------------\n")
    cat("Study design: ")
    cat(paste0(design, " (", type[match(design, designs)], ")\n"))
    cat("log-transformed data (multiplicative model)\n")
    cat(formatC(nsims, format = "d", big.mark = ",", decimal.mark = "."),
        "studies in each iteration simulated.\n\n")
    txt <- paste0("CVwR ", sprintf("%.4g", CVwR))
    if (length(CV) == 2) {
      txt <- paste0(txt, ", CVwT ", sprintf("%.4g", CVwT), ", ")
    } else {
      txt <- paste0(txt, ", ")
    }
    cat(paste0(txt, "n(i) ", paste0(n, collapse = "|"), " (N ", sum(n),
                    ")\n"))
    txt <- paste0("Nominal alpha                 : ", signif(alpha, 5))
    if (!is.na(alpha.pre) && (alpha.pre != alpha)) {
      txt <- paste0(txt, ", pre-specified alpha ", alpha.pre, "\n")
    } else {
      txt <- paste(txt, "\n")
    }
    cat(txt)
    cat("True ratio                    :", sprintf("%.4f", theta0), "\n")
    cat(paste0("Regulatory settings           : ", reg$name, " (",
               method, ")\n"))
    if (flushable) flush.console() # advance console output.
  }
  TIE[1] <- power.scABEL(alpha = al, CV = CV, theta0 = U, n = n,
                         design = design, regulator = reg,
                         nsims = nsims, setseed = setseed)
  no <- no + nsims
  pwr[1] <- power.scABEL(alpha = al, CV = CV, theta0 = theta0,
                         n = n, design = design, regulator = reg,
                         setseed = setseed)
  no <- no + 1e5
  if (TIE[1] > alpha) { # adjust only if needed (> nominal alpha)
    x         <- uniroot(opt, interval = c(0, alpha), tol = tol)
    alpha.adj <- x$root
    TIE[2] <- power.scABEL(alpha = alpha.adj, CV = CV, theta0 = U, n = n,
                           design = design, regulator = reg,
                           nsims = nsims, setseed = setseed)
    pwr[2] <- power.scABEL(alpha = alpha.adj, CV = CV, theta0 = theta0,
                           n = n, design = design, regulator = reg,
                           setseed = setseed)
  }
  if (!is.na(alpha.adj)) no <- no + nsims*x$iter
  if (details) run.time <- proc.time() - ptm
  if (print) { # fetch and print results
    txt <- paste0("Empiric TIE for alpha ", sprintf("%.4f", al), "  : ",
                  sprintf("%.5f", TIE[1]))
    if (TIE[1] > alpha || alpha.pre != alpha) {
      rel.change <- 100*(TIE[1] - alpha)/alpha
      if (details) {
        txt <- paste0(txt, " (rel. change of risk: ",
                      sprintf("%+1.3g%%", rel.change), ")")
      }
    }
    if (!is.na(pwr[1])) {
      pwr.unadj <- pwr[1]
      txt <- paste0(txt, "\nPower for theta0 ", sprintf("%.4f", theta0),
                         "       : ", sprintf("%.3f", pwr.unadj))
    }
    if (TIE[1] > alpha) {
      if (alpha.adj >= 0.01) {
        txt <- paste0(txt, "\nIteratively adjusted alpha    : ",
                      sprintf("%.5f", alpha.adj),
                      "\nEmpiric TIE for adjusted alpha: ",
                      sprintf("%.5f", TIE[2]))
      } else {
        txt <- paste0(txt, "\nIteratively adjusted alpha    : ",
                      signif(alpha.adj, 3),
                      "\nEmpiric TIE for adjusted alpha: ",
                      sprintf("%.4f", TIE[2]))
      }
    if (!is.na(pwr[2])) {
      pwr.adj <- pwr[2]
      txt <- paste0(txt, "\nPower for theta0 ",sprintf("%.4f", theta0),
                         "       : ", sprintf("%.3f", pwr.adj))
      if (details) {
        txt <- paste0(txt, " (rel. impact: ", sprintf("%+1.3g%%",
                      100*(pwr[2] - pwr[1])/pwr[1]), ")")
      }
      txt <- paste(txt, "\n\n")
    } else {
      txt <- paste0(txt, "\n\n")
    }
      if (details) {
        txt <- paste0(txt, "Runtime    : ", signif(run.time[3], 3),
                      " seconds\nSimulations: ",
                      formatC(no, format = "d", big.mark = ",",
                        decimal.mark = "."), " (",
                      (no - no %% nsims - nsims) / nsims,
                      " iterations)\n\n")
      }
    } else {
      txt <- paste0(txt, "\nTIE not > nominal alpha; ")
      ifelse(alpha.pre == alpha,
        txt <- paste0(txt, "no adjustment of alpha is required.\n\n"),
        txt <- paste0(txt, "the chosen pre-specified alpha is ",
                      "justified.\n\n"))
    }
  cat(txt)
  } else { # print=FALSE
    # Prepare and return list of results.
    res <- list(regulator = reg$name, method = method, design = design,
                type = type[match(design, designs)], alpha = alpha,
                alpha.pre = alpha.pre, CV = CV, n = sum(n), theta0 = theta0,
                TIE.unadj = signif(TIE[1], 5),
                rel.change = ifelse(!is.na(alpha.adj),
                                  signif(100*(TIE[1] - alpha)/alpha, 5), NA),
                pwr.unadj = signif(pwr[1], 5), alpha.adj = signif(alpha.adj, 5),
                TIE.adj = signif(TIE[2], 5), pwr.adj = signif(pwr[2], 5),
                rel.loss = ifelse(!is.na(pwr[2]),
                                  signif(100*(pwr[2] - pwr[1])/pwr[1], 5), NA),
                sims = no)
    return(res)
  }
}
# Examples
# 1. using all defaults
#   scABEL.ad()
# should return:
#   +++++++++++ scaled (widened) ABEL +++++++++++
#            iteratively adjusted alpha
#   ---------------------------------------------
#   Study design: 2x3x3 (RRT|RTR|TRR)
#   log-transformed data (multiplicative model)
#   1,000,000 studies in each iteration simulated.
#
#   CVwR 0.3, n(i) 18|18|18 (N 54)
#   Nominal alpha                 : 0.05
#   True ratio                    : 0.900
#   Regulatory settings           : EMA (ABE)
#   Empiric TIE for alpha 0.0500  : 0.07189
#   Iteratively adjusted alpha    : 0.03389
#   Empiric TIE for adjusted alpha: 0.05000
#   Power for theta0 0.900        : 0.764
#
# 2. Explore the impact on power.
#   scABEL.ad(regulator="EMA", design="2x2x4", CV=0.3, n=34, details=TRUE, theta0=0.9)
# should return:
#   +++++++++++ scaled (widened) ABEL +++++++++++
#            iteratively adjusted alpha
#   ---------------------------------------------
#   Study design: 2x2x4 (RTRT|TRTR)
#   log-transformed data (multiplicative model)
#   1,000,000 studies in each iteration simulated.
#
#   CVwR 0.3, n(i) 17|17 (N 34)
#   Nominal alpha                 : 0.05
#   True ratio                    : 0.900
#   Regulatory settings           : EMA (ABE)
#   Empiric TIE for alpha 0.0500  : 0.08163 (rel. change of risk: +63.3%)
#   Power for theta0 0.900        : 0.803
#   Iteratively adjusted alpha    : 0.02857
#   Empiric TIE for adjusted alpha: 0.05000
#   Power for theta0 0.900        : 0.725 (rel. impact: -9.68%)
#
#   Runtime    : 5.66 seconds
#   Simulations: 5,100,000 (4 iterations)
#
# 3. Explore whether a pre-specified alpha maintains the consumer's risk;
#    different CVs of T and R and slight unbalance in the data expected.
#   scABEL.ad(regulator="EMA", design="2x2x4", CV=CVp2CV(0.3, ratio=0.75), n=c(15, 14), alpha.pre=0.025)
# should return:
#   +++++++++++ scaled (widened) ABEL +++++++++++
#            iteratively adjusted alpha
#   ---------------------------------------------
#   Study design: 2x2x4 (RTRT|TRTR)
#   log-transformed data (multiplicative model)
#   1,000,000 studies in each iteration simulated.
#
#   CVwR 0.3217, CVwT 0.2769, n(i) 15|14 (N 29)
#   Nominal alpha                 : 0.05, pre-specified alpha 0.025
#   True ratio                    : 0.900
#   Regulatory settings           : EMA (ABEL)
#   Empiric TIE for alpha 0.0250  : 0.03941
#   TIE not > nominal alpha; the chosen pre-specified alpha is justified.
#
# 4. Assign to a variable and subsequently call the results
#   x <- scABEL.ad(regulator="EMA", design="2x2x4", CV=0.3, n=34, theta0=0.9, print=FALSE, alpha.pre=0.025)
# TIE for the requested alpha
#   x$TIE.unadj
#   [1] 0.044427 # No inflation of the TIE; 0.025 is justified.
#   x$pwr.unadj
#   [1] 0.70537  # Consider to increase the sample size!
